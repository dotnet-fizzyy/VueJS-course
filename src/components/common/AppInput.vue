<template>
    <div :class="$style.root">
        <span v-if="!!label" :class="$style.label">{{ label }}</span>

        <input :class="$style.input" v-font-size="fontSize" :placeholder="placeholder" :value="value" @input="change" />

        <div v-if="!!errorMessage" :class="$style['error-message-container']">
            <error-icon />
            <span :class="$style['error-message']">{{ errorMessage }}</span>
        </div>
    </div>
</template>

<script lang="ts">
import ErrorIcon from '@/components/icons/ErrorIcon.vue';
import Vue from 'vue';
import { DefaultFontSize } from '@/constants/styles';
import { FontSizeDirective, FontSizeDirectiveName } from '@/directives/fontSizeDirective';

export interface AppInputProps {
    value: string;
    label?: string;
    placeholder?: string;
    errorMessage?: string;
    fontSize?: number;
    change: (value: string) => void;
}

export default Vue.extend({
    name: 'AppInput',
    components: { ErrorIcon },
    directives: {
        [FontSizeDirectiveName]: FontSizeDirective,
    },
    props: {
        label: {
            type: String,
        },
        placeholder: {
            type: String,
        },
        value: {
            type: String,
            required: true,
        },
        fontSize: {
            type: Number,
            default: DefaultFontSize,
        },
        errorMessage: {
            type: String,
        },
    },
    methods: {
        change(event): void {
            const value: string = event.target.value;

            this.$emit('change', value);
        },
    },
});
</script>

<style lang="scss" module scoped>
@import '../../styles/inputs';
@import '../../styles/fonts';

.root {
    width: 100%;
    height: 100%;
    display: flex;
    flex-direction: column;
}

.label {
    @extend .font;

    color: getRgbaWhite($white-medium-opacity);
    margin-bottom: 15px;
}

.error-message-container {
    display: flex;
    flex-direction: row;
    align-items: center;
    margin-top: 7px;
}

.error-message {
    color: $light-red;
    margin-left: 5px;
    font-family: $roboto;
    font-size: 14px;
}
</style>
