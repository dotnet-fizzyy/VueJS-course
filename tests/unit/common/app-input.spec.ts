import AppInput from '@/components/common/AppInput.vue';
import { shallowMount } from '@vue/test-utils';

describe('AppInput.vue Tests', () => {
    test('Should render', () => {
        //Arrange
        const wrapper = shallowMount(AppInput, {
            propsData: {
                label: 'Label',
                value: 'Value',
                placeholder: 'Placeholder',
                errorMessage: 'Error message',
                fontSize: 16,
            },
        });

        //Act & Assert
        expect(wrapper.element).toMatchSnapshot();
    });

    test('Should not render "label" block if it was not provided via props', () => {
        //Arrange
        const wrapper = shallowMount(AppInput, {
            propsData: {
                value: 'Value',
                placeholder: 'Placeholder',
                errorMessage: 'Error message',
                fontSize: 16,
            },
        });

        //Act & Assert
        expect(wrapper.element).toMatchSnapshot();
    });

    test('Should not render "error" block if it was not provided via props', () => {
        //Arrange
        const wrapper = shallowMount(AppInput, {
            propsData: {
                label: 'Label',
                value: 'Value',
                placeholder: 'Placeholder',
                fontSize: 16,
            },
        });

        //Act & Assert
        expect(wrapper.element).toMatchSnapshot();
    });

    test('Should emit event with changed value in input', () => {
        //Arrange
        const wrapper = shallowMount(AppInput, {
            propsData: {
                label: 'Label',
                value: '',
                placeholder: 'Placeholder',
                errorMessage: 'Error message',
                fontSize: 16,
            },
        });

        const expectedValue = 'Hello world!';

        //Act & Assert
        const inputWrapper = wrapper.find('input');

        // @ts-ignore Value exists in `element`
        expect(inputWrapper.element.value).toEqual('');

        inputWrapper.setValue(expectedValue);

        expect(wrapper.emitted().change).toBeTruthy();
        expect(wrapper.emitted().change[0][0]).toEqual(expectedValue);
    });
});
