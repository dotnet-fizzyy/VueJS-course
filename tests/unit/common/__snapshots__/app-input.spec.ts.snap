// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppInput.vue Tests Should not render "error" block if it was not provided via props 1`] = `
<div
  class="root"
>
  <span
    class="label"
  >
    Label
  </span>
   
  <input
    class="input"
    placeholder="Placeholder"
    style="font-size: 16px;"
  />
   
  <!---->
</div>
`;

exports[`AppInput.vue Tests Should not render "label" block if it was not provided via props 1`] = `
<div
  class="root"
>
  <!---->
   
  <input
    class="input"
    placeholder="Placeholder"
    style="font-size: 16px;"
  />
   
  <div
    class="error-message-container"
  >
    <error-icon-stub />
     
    <span
      class="error-message"
    >
      Error message
    </span>
  </div>
</div>
`;

exports[`AppInput.vue Tests Should render 1`] = `
<div
  class="root"
>
  <span
    class="label"
  >
    Label
  </span>
   
  <input
    class="input"
    placeholder="Placeholder"
    style="font-size: 16px;"
  />
   
  <div
    class="error-message-container"
  >
    <error-icon-stub />
     
    <span
      class="error-message"
    >
      Error message
    </span>
  </div>
</div>
`;
